mlflow/tracing/export/mlflow_v3.py

```diff
@@ -176,7 +176,37 @@ def _log_trace(self, trace: Trace, prompts: Sequence[PromptVersion]) -> None:
             if trace:
                 add_size_stats_to_trace_metadata(trace)
                 returned_trace_info = self._client.start_trace(trace.info)
+
+                # Collect attachments from all spans and update their references
+                attachments = []
+                attachment_ref_mapping = {}  # old_ref -> new_ref
+
+                for span in trace.data.spans:
+                    if hasattr(span, "_attachments") and span._attachments:
+                        # Update attachment references with real artifact_uri
+                        from mlflow.tracing.utils.artifact_utils import get_artifact_uri_for_trace
+
+                        artifact_uri = get_artifact_uri_for_trace(returned_trace_info)
+
+                        updated_attachments = {}
+                        for old_ref, attachment in span._attachments.items():
+                            # Generate new reference with real artifact_uri
+                            new_ref = attachment.ref(trace.info.trace_id, artifact_uri)
+                            updated_attachments[new_ref] = attachment
+                            attachment_ref_mapping[old_ref] = new_ref
+                            attachments.append(attachment)
+
+                        # Update the span's attachment references
+                        span._attachments = updated_attachments
+
+                # Update attachment references in span inputs and outputs
+                self._update_attachment_refs_in_spans(trace.data.spans, attachment_ref_mapping)
+
+                # Now upload the trace data with updated references
                 self._client._upload_trace_data(returned_trace_info, trace.data)
+
+                if attachments:
+                    self._client._upload_attachments(returned_trace_info, attachments)
```

Comment:
Could we fetch attachments with trace_manager?

Reply:
For example we can save mapping between trace_id -> attachments when registering spans in the trace_manager, and when exporting we pop the attachments by trace_id and upload them. So you don't need to iterate over spans here to fetch attachments

Reply:
Great suggestion. I'll explore that option too.